examples/sparse/hgnn.py

```diff
@@ -0,0 +1,95 @@
+"""
+Hypergraph Neural Networks (https://arxiv.org/pdf/1809.09401.pdf)
+"""
+import dgl
+import dgl.data
+import dgl.mock_sparse
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from torchmetrics.functional import accuracy
+import tqdm
+
+dataset = dgl.data.CoraGraphDataset()
+
+graph = dataset[0]
+# The paper created the hypergraph by "each time one vertex in the
+# graph is selected as the centroid and its connected vertices
+# are used to generate one hyperedge including the centroid
+# itself".  In this case, the incidence matrix of the hypergraph
+# is the same as the adjacency matrix of the original graph (with
+# self-loops).
+# We follow the paper assuming that the rows of the incidence matrix
+# are nodes and the columns are edges.
+
+# QUESTION: can we either (1) easily convert a PyTorch sparse matrix
+# to a DGL sparse matrix, or (2) make g.adj() a DGL sparse matrix
+# in the future?
+adj = dgl.add_self_loop(graph).adj().coalesce()
+row, col = adj.indices()
+values = adj.values()
+H = dgl.mock_sparse.create_from_coo(row, col, values)
+
+X = graph.ndata["feat"]
+Y = graph.ndata["label"]
+train_mask = graph.ndata["train_mask"]
+val_mask = graph.ndata["val_mask"]
+test_mask = graph.ndata["test_mask"]
+
+
+class HGNN(nn.Module):
+    def __init__(self, in_size, out_size, hidden_dims=16):
+        super().__init__()
+
+        self.Theta1 = nn.Linear(in_size, hidden_dims)
+        self.Theta2 = nn.Linear(hidden_dims, out_size)
+        self.dropout = nn.Dropout(0.5)
+
+    def forward(self, H, X):
+        d_V = H.sum(1)  # node degree
+        d_E = H.sum(0)  # edge degree
+        D_V_invsqrt = dgl.mock_sparse.diag(d_V ** -0.5)  # D_V ** (-1/2)
+        D_E_inv = dgl.mock_sparse.diag(d_E ** -1)  # D_E ** (-1)
+        # Minor suggestion: could we have dgl.mock_sparse.eye(N)?
+        W = dgl.mock_sparse.diag(torch.ones(d_E.shape[0]))
+
+        conv = D_V_invsqrt @ H @ W @ D_E_inv @ H.T @ D_V_invsqrt
```

Comment:
L49-56 can be done in the pre-processing step.

Reply:
If `H` is an argument in the forward function then technically I cannot put it in the preprocessing step.  So I guess the problem is whether `H` should be in the forward function or in `__init__`.